### DISCUSSION TITLE
Using a URL to fetch a design from anywhere.

### DISCUSSION THREAD

**Discussion by gbishop** [UNC-Project-Open-AAC/OS-DPI#238] [Announcements]
I'm working on code to make it possible to give someone a URL they can click to both fetch and display a design (like our demo links) except the link can point to anywhere publically accessible. As an extra feature I want to make it possible for you to update the design in place and give the user a button (or whatever) in the UI allowing them to update to the latest without using the designer interface.

I'll update this post later when it is working. The point of this page is to give me a place to plant a design for testing.

If there are multiple matches it takes the first one. I can swap to two links below to change which one get fetched.

[test_design.zip](https://github.com/UNC-Project-Open-AAC/OS-DPI/files/13923736/test_design.zip)

[test_design.zip](https://github.com/UNC-Project-Open-AAC/OS-DPI/files/13923745/test_design.zip)

I can include multiple designs in the same page by using the hash on the end of the URL as shown below.

[another_test_design.zip](https://github.com/UNC-Project-Open-AAC/OS-DPI/files/13923818/another_test_design.zip)

I've left the links visible below but they could be have text instead of the full URL

http://localhost:8080/OS-DPI/?fetch=https://github.com/UNC-Project-Open-AAC/OS-DPI/discussions/238#test_design

[Try the test_design](http://localhost:8080/OS-DPI/?fetch=https://github.com/UNC-Project-Open-AAC/OS-DPI/discussions/238#test_design)

http://localhost:8080/OS-DPI/?fetch=https://github.com/UNC-Project-Open-AAC/OS-DPI/discussions/238#another_test_design

[Try another design](http://localhost:8080/OS-DPI/?fetch=https://github.com/UNC-Project-Open-AAC/OS-DPI/discussions/238#another_test_design)

**Comment by sofia-bg** [comment]
Are you still working on the "try it here" link? Currently does not work for me - takes me to a "site cannot be reached" page

**Reply by gbishop** [reply]
Yes. It is pointing to the version running on my home computer. It won't
work for anyone else.

**Comment by gbishop** [comment]
OK. I've got a crazy idea that could make it work. If the URL you provide is not suffixed with ".osdpi" or ".zip" then it will attempt to fetch it as HTML and search that text for a link with one of those suffixes. This would not allow for multiple uploads on the same page. Maybe another query parameter specifies the name? That could work. The URLs will look terrible but no one should be reading (much less typing) URLs. You could always use a URL shortener. 

So the URL might look like:

`https://unc-project-open-aac.github.io/OS-DPI/?fetch=https://github.com/UNC-Project-Open-AAC/OS-DPI/discussions/238&name=test_design.zip`

Then we could provide some shortcuts so it could become:

`https://unc-project-open-aac.github.io/OS-DPI/?fetch=OS-DPI/discussions/238&name=test_design.zip`

That level of indirection through the page maybe is sufficient to allow updates.

I'll nap on that.

**Reply by sofia-bg** [reply]
Following up on this - can you walk me through how I make this happen? Do I first need to export my design and save it to GitHub?

**Reply by gbishop** [reply]
Sorry, I've been sitting on this release too long. I just pushed it to production. 

The steps to make your design accessible with a URL on in a Github discussion.

1. Export your design.
2. Change the file extension from `.osdpi` to `.zip`
3. Drag the file into a discussion. Like this: 
[test_design.zip](https://github.com/UNC-Project-Open-AAC/OS-DPI/files/14495854/test_design.zip)
4. Note the URL it created for the file. The URL above is `https://github.com/UNC-Project-Open-AAC/OS-DPI/files/14495854/test_design.zip`
5. The URL to give to others will be https://unc-project-open-aac.github.io/OS-DPI/?fetch=https://github.com/UNC-Project-Open-AAC/OS-DPI/files/14495854/test_design.zip which is `https://unc-project-open-aac.github.io/OS-DPI/?fetch=` prepended to the URL they create.

Or, you can upload the file to Github without changing the file extension. Then browse to the page for the file and right click the `View Raw` link to copy the URL. Prepend the `https://unc-project-open-aac.github.io/OS-DPI/?fetch=` to that URL and it should work. For example: https://unc-project-open-aac.github.io/OS-DPI/?fetch=https://github.com/UNC-Project-Open-AAC/OS-DPI/raw/main/examples/updated/grid_ex_1.osdpi

Or, you *should* be able to upload it anywhere publically accessible, for example a WordPress site and do the same trick. 

This relies on a tiny proxy that is running on my office computer. I was disappointed to see that it has stopped working when I just checked. I clearly have some debugging to do. 

We need a more permanent solution for sure.

**Reply by sofia-bg** [reply]
Thanks that worked for me to post in discussions to see the URL and prepend it. It did not work for me to upload to github and view raw, got the message that it could not view a file so big.

**Reply by gbishop** [reply]
You don't want to click that `View Raw` link. You want to Right Click it and copy the URL. There may be a better way to get the URL of the RAW data but I don't know it.

**Reply by gbishop** [reply]
After Googling a bit I see the preferred format is: `https://raw.githubusercontent.com/UNC-Project-Open-AAC/OS-DPI/master/examples/updated/grid_ex_1.osdpi`

So, if a file is in `folder1/folder2/design.osdpi` the *raw* URL would be: `https://raw.githubusercontent.com/UNC-Project-Open-AAC/OS-DPI/master/` prepended to `folder1/foder2/design.osdpi`

**Comment by aetkie** [comment]
I am trying to follow these steps and have not been successful.

This is the design:
[English_Mandarin_grid.osdpi.zip](https://github.com/UNC-Project-Open-AAC/OS-DPI/files/14672084/English_Mandarin_grid.osdpi.zip)

This is the link: https://github.com/UNC-Project-Open-AAC/OS-DPI/files/14672084/English_Mandarin_grid.osdpi.zip

This is the link prepended: https://unc-project-open-aac.github.io/OS-DPI/?fetch=https://github.com/UNC-Project-Open-AAC/OS-DPI/files/14672074/English_Mandarin_grid.osdpi.zip 

This is what I get when I copy and paste into a new tab: 


<img width="764" alt="Screenshot 2024-03-20 at 4 13 45 PM" src="https://github.com/UNC-Project-Open-AAC/OS-DPI/assets/113473933/cd3c0a12-629a-4a1d-afcd-d12e2fd2f754">

**Comment by gbishop** [comment]
It should be English_Mandarin_grid.osdpi or English_Mandarin_grid.zip but
not English_Mandarin_grid.osdpi.zip as you have it. The .osdpi file is
already a zip file. When you zip it, you then have a zip file inside a zip
file and that can't work.

Simply rename the original file from English_Mandarin_grid.osdpi
to English_Mandarin_grid.zip so that GitHub will let you attach it.

**Reply by aetkie** [reply]
Thanks @gbishop. I fixed the extension. Now the link prepended is https://unc-project-open-aac.github.io/OS-DPI/?fetch=https://github.com/UNC-Project-Open-AAC/OS-DPI/files/14672332/English_Mandarin_grid.zip 
But I am still getting the same error message.

**Comment by gbishop** [comment]
You changed the name but you didn't unzip it. It is still a zip file
containing the .osdpi file.

Go back to the original design file. The .osdpi you exported from the app.
Change the suffix to ".zip" and then use that.

**Comment by aetkie** [comment]
I am still getting the same error message. When I am in the design, I export it. It saves to my desktop. I change the extension to .zip. Is that correct?

**Comment by gbishop** [comment]
Correct. Then attach that to a comment (or whatever) on GitHub. Get that
URL and it should work.

**Comment by aetkie** [comment]
Trying this again because I am still getting the same error message. 

This is the exported design: 
[keyboard.zip](https://github.com/UNC-Project-Open-AAC/OS-DPI/files/14697344/keyboard.zip)

This is the link: https://github.com/UNC-Project-Open-AAC/OS-DPI/files/14697348/keyboard.zip

This is the prepended link: https://unc-project-open-aac.github.io/OS-DPI/?fetch=https://github.com/UNC-Project-Open-AAC/OS-DPI/files/14697348/keyboard.zip

**Comment by gbishop** [comment]
Looks like you have done everything correctly. There must be a bug. I'll
look into it.

**Comment by gbishop** [comment]
How did you upload that file?

**Comment by aetkie** [comment]
I opened a discussion post and dragged the design from my desktop into the box

**Comment by gbishop** [comment]
OK. I see the issue. The proxy has died again... Or more likely it didn't
restart when the machine rebooted.

I just restarted it. That link should work now.

**Reply by aetkie** [reply]
it works! thanks for troubleshooting with me

**Comment by gbishop** [comment]
I'll figure out why the proxy goes away. Maybe I can come up with a better
error message as well. At least knowing the proxy wasn't working would help.
